### PR TITLE
tests passed

### PR DESCRIPTION
**Pull request by invegat**

### CODE REVIEW

File: src/arrays.js [hunk 1]
```diff
@@ -9,37 +9,87 @@ const each = (elements, cb) => {
...
+    cb(elements[i], i);
+  }
 };
 
-const map = (elements, cb) => {
-  // Produces a new array of values by mapping each value in list through a transformation function (iteratee).
-  // Return the new array.
-};
 
 const reduce = (elements, cb, startingValue) => {
```
Comment (SunJieMing) [inline src/arrays.js:18]: You can also do this for a starting value: `(elements, cb, startingValue = elements.shift())` for the default value.  This takes advantage of the ES6 default parameters.

File: src/arrays.js [hunk 2]
```diff
@@ -9,37 +9,87 @@ const each = (elements, cb) => {
...
   // Combine all elements into a single value going from left to right.
   // Elements will be passed one by one into `cb`.
   // `startingValue` is the starting value.  If `startingValue` is undefined then make `elements[0]` the initial value.
+  if (startingValue === undefined) {
+    startingValue = (typeof elements[0]) === 'string' ? '' : 0;
```
Comment (SunJieMing) [inline src/arrays.js]: Just make `startingValue` the first item from the array

File: src/arrays.js (hunk 1 above)
Comment (SunJieMing) [reply src/arrays.js:18]: Unless of course you are avoiding modifying the input array

File: src/arrays.js (hunk 2 above)
Comment (SunJieMing) [reply src/arrays.js]: A reduce is supposed to have the first item in the array as its initial memo

File: src/arrays.js [hunk 3]
```diff
@@ -9,37 +9,87 @@ const each = (elements, cb) => {
...
+  for (let i = 0; i < elements.length; i++) {
+    acc = cb(acc, elements[i], i);
+  }
+  return acc;
+};
+
+const map = (elements, cb) => {
+  // Produces a new array of values by mapping each value in list through a transformation function (iteratee).
+  // Return the new array.
+  return reduce(elements, (acc, v) => {
```
Comment (SunJieMing) [inline src/arrays.js:37]: Really creative

File: src/arrays.js [hunk 4]
```diff
@@ -9,37 +9,87 @@ const each = (elements, cb) => {
...
+      acc.push(v);
+    }
+    return acc;
+  }, []);
 };
 
 /* Extra Credit */
-
+function isArray(elements) {
+  return (Object.prototype.toString.call(elements) === '[object Array]');
```
Comment (SunJieMing) [inline src/arrays.js]: The preferred way of checking whether a variable is an array or not is this: `Array.isArray(myVariable);`

Comment (SunJieMing) [reply src/arrays.js]: Good job creating another function though

File: src/callbacks.js [hunk 5]
```diff
@@ -24,58 +24,77 @@
 // code here
 
 const foods = ['pineapple', 'mango', 'ribeye', 'curry', 'tacos', 'ribeye', 'mango'];
+function firstItem(foods,f) {
```
Comment (SunJieMing) [inline src/callbacks.js]: I'd prefer it if you put a space between the parameters

File: src/callbacks.js [hunk 6]
```diff
@@ -24,58 +24,77 @@
...
+  return f(foods[0])
```
Comment (SunJieMing) [inline src/callbacks.js]: This line should have a semicolon

File: src/callbacks.js [hunk 7]
```diff
@@ -24,58 +24,77 @@
...
 
 firstItem(foods, (firstItem) => {
   console.log(`The first item is ${firstItem}.`);
 });
 
 // Write a function called getLength that passes the length of the array into the callback
 // code here
-
+function getLength(foods,f) {
+  return f(foods.length)
```
Comment (SunJieMing) [inline src/callbacks.js]: Make sure that you stay consistent.  Inconsistent code is a mark of an inexperienced developer.  It shows experience and discipline to keep your code consistent.

File: src/objects.js [hunk 8]
```diff
@@ -1,26 +1,47 @@
...
+const each = require('./arrays.js').each;
 
 const keys = (obj) => {
   // Retrieve all the names of the object's properties.
   // Return the keys as strings in an array.
   // Based on http://underscorejs.org/#keys
+  return Object.keys(obj);
 };
-
+function isFunction(func) {
```
Comment (SunJieMing) [inline src/objects.js:11]: For the sake of consistency this should be written the same way the other functions are written.

File: tests/arrays.test.js [hunk 9]
```diff
@@ -66,7 +66,7 @@ describe('arrays', () => {
       const arr = [1, 2, 3, 4, 5];
       const result = arrayMethods.reduce(arr, callBackMockFn);
       expect(result).toBe(25);
-      expect(callBackMockFn.mock.calls.length).toBe(4);
+      expect(callBackMockFn.mock.calls.length).toBe(5);
```
Comment (SunJieMing) [inline tests/arrays.test.js]: This test should not be modified.  The memo in the reduce should be the first item from the array and so the callback would only be invoked 4 times.

File: src/arrays.js (hunk 3 above)
Comment (invegat) [reply src/arrays.js:37]: thanks

File: src/callbacks.js (hunk 5 above)
Comment (invegat) [reply src/callbacks.js]: accustomed  to  getting a 'comma-spacing' error by the eslint configs I have been working with

**Pull request closed without merging**